File: python-cli/src/convex_notes/handlers/watch.py
"""Watch notes handler (real-time subscription)."""

from rich.console import Console

from convex_notes.service import NoteService
from convex_notes.ui import display_notes

console = Console()


def execute(service: NoteService) -> None:
    """Execute the watch notes command."""
    console.print("\n[bold magenta]👀 Watching notes for real-time updates...[/bold magenta]")
    console.print("[dim]   Press Ctrl+C to stop[/dim]\n")

    try:
        for notes in service.subscribe():
            console.print("[cyan]📨[/cyan] [white]Update received:[/white]")
            if not notes:
                console.print("   [yellow]No notes[/yellow]")
            else:
                display_notes(notes)
            console.print("[dim]" + "─" * 50 + "[/dim]")
    except KeyboardInterrupt:
        console.print("\n[magenta]👋 Stopped watching.[/magenta]")

File: backend/crates/cli/src/handlers/watch.rs
//! Watch notes handler (real-time subscription).

use anyhow::Result;
use colored::Colorize;
use futures::StreamExt;

use db::{NoteRepository, NoteService};

use crate::ui::table;

/// Execute the watch notes command.
pub async fn execute<R: NoteRepository>(service: &NoteService<R>) -> Result<()> {
    println!(
        "\n{}",
        "👀 Watching notes for real-time updates..."
            .bright_magenta()
            .bold()
    );
    println!("{}", "   Press Ctrl+C to stop".bright_black());
    println!();

    let mut stream = service.subscribe().await?;

    while let Some(result) = stream.next().await {
        println!(
            "{} {}",
            "📨".bright_cyan(),
            "Update received:".bright_white()
        );

        match result {
            Ok(notes) => {
                if notes.is_empty() {
                    println!("   {}", "No notes".bright_yellow());
                } else {
                    table::display_notes(&notes);
                }
            }
            Err(e) => println!("   {} {}", "Error:".bright_red(), e),
        }

        println!("{}", "─".repeat(50).bright_black());
    }

    Ok(())
}
